### ISSUE TITLE
Flexible layout templates: Set ID attribute for rows

### ISSUE THREAD

**Issue by olafgrabienski** [backdrop/backdrop-issues#4052]
When I configure a Flexible layout row, I want to add an `id` attribute for the row. Setting an ID allows linking to a row using a URL fragment. Example:

- Row: `<section id="about" ...>`
- Link: `<a href="#about" ...>`

The custom ID would be very helpful building one pager websites or landing pages where navigation links point to different sections of the same page.

At the moment, it's possible to add a *Region name* and an *Additional row class* via the "Configure row" dialog. I suggest to also provide the option for an ID.

**Alternatives that have been considered**
I didn't find a way to set custom IDs via the Flexible layout template or via the configuration file.

**Comment by olafgrabienski** [comment]
Here's a simplified example of my use case -- a home page with some sections, all sections linked from a sticky top navigation.

```
<nav>
  <a href="#intro">Intro</a>
  <a href="#about">About</a>
  <a href="#contact">Contact us</a>
</nav>

<section id="intro>
  ...
</section>

<section id="about>
  ...
</section>

<section id="contact>
  ...
</section>
```

As far as I understand, it's not possible to achieve that with a *Flexible layout*.

**Comment by klonos** [comment]
@olafgrabienski have you tried adding the id by overriding the template? So change [this line here](https://github.com/backdrop/backdrop/blob/1.x/core/modules/layout/templates/layout--flexible.tpl.php#L30), to something like:

```php
<<?php print $row['element']; ?> data-row-id="<?php print $name; ?>" id="<?php print $name; ?>" class="flexible-row <?php print 'l-' . $name; ?>" <?php print $row['row_id']; ?>>
```

Can you please test and see if that works for you?

**Comment by olafgrabienski** [comment]
@klonos Thanks for the hint! I guess, `id="<?php print $name; ?>"` could be a workaround (will test it tomorrow).

In contrast to `<section id="contact">` the result would however be ugly, fictive example:
  `<section id="q84834858-0q840q8we8rqw8er8qw0er-09808r8q0w8er0">`

Using such an ID in a link would result in a barely readable URL in the address bar:
  `https://www.example.com#q84834858-0q840q8we8rqw8er8qw0er-09808r8q0w8er0`

**Comment by klonos** [comment]
True @olafgrabienski ...not ideal.

I am curious what `<?php print $row['row_id']; ?>` in that line does? 🤔

**Comment by olafgrabienski** [comment]
> I guess, `id="<?php print $name; ?>"` could be a workaround (will test it tomorrow)

@klonos, just tested, here's the result, including an ID:

`<div data-row-id="fb6a272b-930c-47cd-b900-f5e80679b157" id="fb6a272b-930c-47cd-b900-f5e80679b157" class="flexible-row l-fb6a272b-930c-47cd-b900-f5e80679b157">`

As expected, the ID is anything but ideal. Apart of that, *Flexible layout templates* were mainly made for people who don't write Layout templates. Keeping my request: Flexible layouts should provide the option to set a custom ID per row via the user interface.

**Comment by jenlampton** [comment]
Are there any other examples of doing this in core? I know we have a lot of places where people can add classes, but an ID might be taking it too far. I don't think we should allow anyone to add an ID via the UI.

**Comment by laryn** [comment]
Haven't tested @olafgrabienski but as a hack in your template override, what would happen using `urlencode($region['region_name'])` for the id instead of `$name`?

**Comment by olafgrabienski** [comment]
@jenlampton Interesting approach to put a name attribute in the content. I guess, IDs are a better fit for my use case, will elaborate it tomorrow.

@laryn also interesting, I'll test it.

**Comment by olafgrabienski** [comment]
Oh, and the `name` attribute doesn't seem to be supported anymore for `<a>` tags in HTML5, cf. https://www.w3schools.com/tags/tag_a.asp.

**Comment by jenlampton** [comment]
Well that explains some things (and dates me!) :)

**Comment by olafgrabienski** [comment]
On yesterday's weekly meeting I was asked to explain my use case more in detail. For now, I can't post a screenshot of the design I'm currently building but I can provide a similar example:

https://blackrockdigital.github.io/startbootstrap-freelancer (Screenshot further below)

When you look at the HTML, the navigation items point to sections with IDs. As far as I know, nearly all one page websites with navigation menu work this way. They need the IDs for navigation.

IDs per section are also helpful for other things, e.g. as a reference for scrollspy behavior via waypoints. The code examples I found are using sections with IDs, I guess because it's common to build such pages this way.

Finally, I don't understand the objections against an option for IDs in the user interface, at least in this case. I guess, Layouts are mainly built by site architects. Most of them are able to write own layout templates in code and add IDs without restrictions. Now, as Flexible layouts provide an interface for Layout templates, what's the problem with IDs? In this case, the IDs wouldn't be used by content editors who might not know how to handle them (e.g. unique IDs on a page).

Hm, I guess I'm stating the obvious, or is there something I didn't consider?

---

**Screenshot of the above mentioned example:**

![screen-backdrop-flexible-layout-section-ids-example-bootstrap-freelancer](https://user-images.githubusercontent.com/14188307/65328436-9c0ae000-dbb6-11e9-8af0-8c128b9467fa.png)

**Comment by laryn** [comment]
I agree that the single long-scroll page with anchors is a fairly common need, and beyond that, linking to a section of a page in general can be needed. I'm personally in favor of allowing IDs added to flexible layout rows in the UI.

**Comment by stpaultim** [comment]
@olafgrabienski - For what it's worth, I understood the objection to be that we have not provided the option for adding ID's to anything else and we should be cautious about adding another option to the UI that did not previously exist. 

It was my impression that you did a pretty good job of refuting this concern and that by the end of the discussion there was less of an objection and more of a mild concern. 

Personally, I am of the opinion that the use case is important and that we should provide some option. I was simply not ready to endorse one specific solution until I had a little time to think about it and whether or not there might be an alternative solution that is better. 

But, I definitely want the same thing for my users that you want for yours. I'm just thinking about the best way to provide that for them and it may in fact be your proposed solution. I certainly don't object to your proposal, I'm just trying to wrap my head around the whole thing.

**Comment by jenlampton** [comment]
@olafgrabienski you had a great point that MY OWN SITE works this way :) 

I think the use-case is valid enough to warrant allowing IDs here, and not adding IDs everywhere else. I was most worried because that can be a slippery slope. Consider my objection mostly lifted :)

**Comment by olafgrabienski** [comment]
Hey all, thanks for your feedback! Seems that we can start to consider *how* to allow the input of IDs. As usual, I won't be able to provide the code but I hope I can help otherwise.

One idea: let's mention in a help text that IDs should be unique.

**Comment by klonos** [comment]
I am interested in working on this.

Just a thought before actual implementation: the way things are currently implemented, the only place I see as easy to add the input for IDs is the "Configure row" dialog - not possible to be added to the region configuration (not in the edit form for flexible templates anyway). Putting the input in the "Configure region" dialog ("Manage blocks" form) would make IDs available to regions, even in layouts not using flexible templates. It is my understanding from the above that the IDs should only be limited to flexible templates. Please let me know if otherwise, and I'll add the input in the "Configure region" dialog, with a check to see if the template being used is flexible.

**Comment by jenlampton** [comment]
> the best place I see as easy to add the input for IDs is the "Configure row" dialog

Yep, that sounds great :) Maybe copy how Blocks let you add classes, and do the same?

**Comment by klonos** [comment]
> Maybe copy how Blocks let you add classes (in a collapsed fieldset), and do the same?

Yup, exactly what I was thinking 👍

**Comment by olafgrabienski** [comment]
> what would happen using `urlencode($region['region_name'])` for the id instead of `$name`?

@laryn Probably don't need it anymore but I was curious: I get IDs like `Header`, `News`, `Content`, or `Contact+info`. The latter is from the Region name *Contact info*.

**Comment by olafgrabienski** [comment]
> I am interested in working on this.

@klonos Are you (or again) interested to make a PR to allow a custom ID for Flexibel layout rows? I just rediscovered this issue after reading a related [Forum post](https://forum.backdropcms.org/forum/why-were-all-handy-classes-removed).

**Comment by herbdool** [comment]
I see that this is about ID, but I think that forum post was about classes not being available in flexible layouts. I took a look. We can add classes to rows (actually the container) in the layout but not the region name. Maybe it would be simple enough to create a machine name version of the region name and put that into the classes. It also seems to hardcode the top row as a `<header>` and the bottom row as a `<footer>` (the parents of the containers). Maybe that's not a big deal. But the middle container parent doesn't get anything other than the uuid.

Apologies if this has already been rehashed above.

**Comment by olafgrabienski** [comment]
You're right @herbdool, the Forum post was mainly about classes. So, only slightly related. (An option for a custom row ID might however also be helpful when you're looking for custom classes.)

Providing classes from a machine name of the region name sounds interesting, but regarding the feature request discussed here a class wouldn't help, instead we need indeed a custom ID (as an anchor for navigation purposes).

**Comment by herbdool** [comment]
@olafgrabienski I guess we need a new issue for the classes?

**Comment by olafgrabienski** [comment]
I saw you created it, great! Link: https://github.com/backdrop/backdrop-issues/issues/5032

**Comment by olafgrabienski** [comment]
Based on the comments https://github.com/backdrop/backdrop-issues/issues/4052#issuecomment-533566514 and below I've removed the *needs feedback* label.

@klonos are you still interested to work on this?

**Comment by stpaultim** [comment]
I think the two issues about adding classes to flexible templates are closely related to this issue. Maybe we can get a batch of improvements to flexible layout templates in this next release.

https://github.com/backdrop/backdrop-issues/issues/5032
https://github.com/backdrop/backdrop-issues/issues/5044

**Comment by klonos** [comment]
@olafgrabienski yes, I'm still interested, but:
a) I'm advocating for and working on other issues currently
b) other people are working on PRs that are touching the same files, so I think to best wait for these PRs to be merged first (so that we don't conflict)

**Comment by olafgrabienski** [comment]
@klonos I guess we lost sight of this issue, custom IDs in flexible layout rows. Are you still interested in working on this or would you prefer if I ask others?
